python/ray/experimental/data/datasource.py

```diff
@@ -0,0 +1,216 @@
+import builtins
+from typing import Generic, List, Callable, Union, TypeVar
+
+import ray
+from ray.experimental.data.impl.arrow_block import ArrowRow, \
+    DelegatingArrowBlockBuilder
+from ray.experimental.data.impl.block import Block, ListBlock
+from ray.experimental.data.impl.block_list import BlockList, BlockMetadata
+
+T = TypeVar("T")
+R = TypeVar("WriteResult")
+
+
+class Datasource(Generic[T]):
+    """Interface for defining a custom ``ray.data.Dataset`` datasource.
+
+    To read a datasource into a dataset, use ``ray.data.read_datasource()``.
+    To write to a writable datasource, use ``Dataset.write_datasource()``.
+
+    See ``RangeDatasource`` and ``DummyOutputDatasource`` below for examples
+    of how to implement readable and writable datasources.
+    """
+
+    def prepare_read(self, parallelism: int,
```

Comment:
The datasource can feel free to ignore it. It's not optional since the only caller of this API is dataset, and we always provide it.

Reply:
>It's not optional since the only caller of this API is dataset, and we always provide it.

I'm saying that the dataset shouldn't provide a non-`None` parallelism by default, so if the user doesn't explicitly provide a `parallelism` argument, the data source is free to determine a sensible default. This is a pretty common pattern of deferring the setting of a default config to the component that has the most information for making the best decision for that default.

>The datasource can feel free to ignore it.

I think that sensible defaults determined by the data source that are overridable by the user is better API semantics than a blanket default parallelism of 200 provided by the dataset that the data source can then ignore, since the data source won't be able to distinguish between the dataset default of 200 and a user-provided value.